### DISCUSSION TITLE
create multiple records

### DISCUSSION THREAD

**Discussion by mapsgeek** [pocketbase/pocketbase#853] [Q&A] [answered]
How can i create multiple records in one action?
i'm trying to migrate my current json data to the db, so i tried doing the following
`
array.map(async record=> {
 await client.records.create('groups', record);
})
`
but only one record gets through and then i get a loop of error `Aborted by user`

**Comment by ganigeorgiev** [comment, marked as answer]
This is due to autocancellation. Although you are using `await`, it doesn't actually await anything in your case since you are executing create in a closure inside the map, aka. you are sending the requests for each item of your array in parallel returning an empty `Promise` for each iteration.

You can disable auto cancellation by setting the special `$autoCancel` query parameter to `false`, eg:

```js
client.records.create('groups', record, { '$autoCancel': false });
```

You can learn more about the autocancellation in https://github.com/pocketbase/js-sdk#auto-cancellation.

_In the next release there will be a global helper to disable this behavior for all client requests._

**Reply by mapsgeek** [reply]
I thought that using await would make the executing actually await for the previous one to get done.
i tried wrapping the `create `in a setinterval function and wrapping in a `promis.all` function.
For education purpose can you please show me an example of how to do this in a "right" way in promise based, i mean without using the `autoCanel `prop?

**Reply by ganigeorgiev** [reply]
@mapsgeek Here is an example with both approaches

```js
// insert each record one by one ensuring also the order of the insertion (this may also prevent SQLITE_BUSY/LOCKED errors)
async function importRecordsOneByOne(records) {
    for (let record of records) {
        await client.records.create("example", record);
    }
}

// bulk insert all requests in parallel (insertion order is not guaranteed!)
function importRecordsInParallel(records) {
    const promises = [];
    
    for (let record of records) {
        const promise = client.records.create("example", record, {
            '$autoCancel': false,
        });

        promises.push(promise);
    }

    return Promise.all(promises);
}
```

**Reply by mapsgeek** [reply]
thank you, it seems like i need to understand the difference between .map and for..of better to make sense why one would work and the other don't.

**Reply by ganigeorgiev** [reply]
@mapsgeek  `map` returns a new array where each item is the return result of your callback. Here is an `importRecordsInParallel` equivalent example using `map`:

```js
function importRecordsInParallel(records) {
    const promises = records.map((record) => {
        return client.records.create("example", record, {
            '$autoCancel': false,
        });
    })

    return Promise.all(promises);
}
```

**Reply by ogrotten** [reply]
@ganigeorgiev Thanks for the examples and clarity here.

Is there a method/transaction of some sort to do this as a single huge `CREATE` (like an sql `INSERT`) as opposed to a bunch of little bitty individuals? Is the golang `Dao.RunInTransaction()` exposed for use with the js-sdk?

**Reply by ganigeorgiev** [reply]
@ogrotten Currently transactional create/update/delete is not supported in the web APIs. It is in my todo and will be handled as part of [#48](https://github.com/pocketbase/pocketbase/issues/48#issuecomment-1211775627).

**Reply by ogrotten** [reply]
# 👍

**Comment by jgb-solutions** [comment]
Any updates on this? I'm seeding a PocketBase database right and I'm worried about inserted the data in a loop like that. And is there a way to import existing data from one PB instance to another? @ganigeorgiev

**Comment by andrejbroncek** [comment]
> Any updates on this? I'm seeding a PocketBase database right and I'm worried about inserted the data in a loop like that. And is there a way to import existing data from one PB instance to another? @ganigeorgiev

+1 on this, just spend a while trying to figure out why my array of objects is not being processed.

**Reply by ganigeorgiev** [reply]
> my array of objects is not being processed

@andrejbroncek Please open a new discussion with more details about your problem (if it is related to the default JS SDK autocancellation, please check [Auto cancellation](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#auto-cancellation)).

**Comment by ganigeorgiev** [comment]
I'm closing this discussion for now. Bulk/transactional APIs is in the roadmap but remains a low priority.
